File: solutions/python/two-pointers/longest-mountain-in-array.py
# LeetCode 845 — Longest Mountain in Array (Medium)
# Category: Two Pointers · Approach: Expand From Peaks
# Time: O(n) | Space: O(1)
# Source: https://leetcode.com/problems/longest-mountain-in-array/

def longest_mountain(arr: list[int]) -> int:
    n = len(arr)
    best = 0
    i = 1
    while i < n - 1:
        if arr[i - 1] < arr[i] > arr[i + 1]:
            left = i - 1
            while left > 0 and arr[left - 1] < arr[left]:
                left -= 1
            right = i + 1
            while right < n - 1 and arr[right] > arr[right + 1]:
                right += 1
            best = max(best, right - left + 1)
            i = right + 1
        else:
            i += 1
    return best

File: solutions/typescript/two-pointers/longest-mountain-in-array.ts
// LeetCode 845 — Longest Mountain in Array (Medium)
// Category: Two Pointers · Approach: Expand From Peaks
// Time: O(n) | Space: O(1)
// Source: https://leetcode.com/problems/longest-mountain-in-array/

function longestMountain(arr: number[]): number {
  const n = arr.length;
  let best = 0;
  let i = 1;
  while (i < n - 1) {
    if (arr[i - 1] < arr[i] && arr[i] > arr[i + 1]) {
      let left = i - 1;
      while (left > 0 && arr[left - 1] < arr[left]) left--;
      let right = i + 1;
      while (right < n - 1 && arr[right] > arr[right + 1]) right++;
      best = Math.max(best, right - left + 1);
      i = right + 1;
    } else {
      i++;
    }
  }
  return best;
}
